rust/tools/authorization-logic/src/compilation_top_level.rs

```diff
@@ -33,6 +33,41 @@ use crate::{
 
 use std::fs;
 
+fn merge_multiprogram(multiprog: Vec<AstProgram>) -> AstProgram {
+    AstProgram {
+        relation_declarations: multiprog
+            .iter()
+            .map(|prog| prog.relation_declarations.clone())
+            .flatten()
+            .collect(),
+        assertions: multiprog
+            .iter()
+            .map(|prog| prog.assertions.clone())
+            .flatten()
+            .collect(),
+        queries: multiprog
+            .iter()
+            .map(|prog| prog.queries.clone())
+            .flatten()
+            .collect(),
+        imports: multiprog
+            .iter()
+            .map(|prog| prog.imports.clone())
+            .flatten()
+            .collect(),
+        priv_binds: multiprog
+            .iter()
+            .map(|prog| prog.priv_binds.clone())
+            .flatten()
+            .collect(),
+        pub_binds: multiprog
+            .iter()
+            .map(|prog| prog.pub_binds.clone())
+            .flatten()
+            .collect(),
+    }
+}
```

Comment:
This looks like it could be more naturally expressed as a `fold` upon the `multiprog` vector, where the `fold` lambda just merges the contents of each subcollection of two `AstProgram`s, producing one resultant `AstProgram`.

Reply:
Defining this with `fold` would require first defining an `AstProgram -> AstProgram -> AstProgram` function, and that function would look pretty similar to this one, so I think it is better to leave this as is.

Reply:
You could do it entirely in a lambda in the fold call. It would look something like this:

```Rust
multiprog.iter().fold(None, |acc, currentAstProg|
  Some(match acc {
    None => currentAstProg
    Some(combinedAstProg) => AstProgram {
      acc.relation_declarations.append(currentAstProg.relation_declarations);
      acc.assertions.append(currentAstProg.assertions);
      acc.queries.append(currentAstProg.queries);
      acc.imports.append(currentAstProg.imports);
      acc.priv_binds.append(currentAstProg.priv_binds);
      acc.pub_binds.append(currentAstProg.pub_binds);
      acc
    })
  }).collect().unwrap(); // At least one program required.
```

The benefit of the above is that you don't have to perform the collection iteration stuff for each array, which makes it more readable.

Reply:
fixed